src/Compilers/CSharp/Portable/BoundTree/LengthBasedStringSwitchData.cs

```diff
@@ -0,0 +1,313 @@
+    // then bucketing on a character position selected by heuristic
+    // and finally switching to exact string (this is a simple string comparison when only one possibility remains).
+    //
+    // The benefit of this approach is that it much reduces the need for computing
+    // the input string's hashcode.
+    //
+    // We emit something like:
+    //
+    // // null case:
+    // if (key is null)
+    //   goto labelNull; OR goto labelDefault;
+    //
+    // switch (key.Length)
+    // {
+    //   // empty string doesn't need a char or string test
+    //   case 0: goto labelEmpty;
+    //
+    //   // strings of length 1 don't need any further validation once we've checked one char
+    //   case 1:
+    //     switch (key[posM])
+    //     {
+    //        case '1': goto label1;
+    //        case '2': goto label2;
+    //        ...
+    //        default: goto labelDefault;
+    //     }
+    //   ...
+    //   // when a given length is sufficient to narrow down to one case we skip the char test:
+    //   case N: if (key == "caseN") { goto labelN; } else { goto labelDefault; }
+    //   ...
+    //   case M:
+    //     switch (key[posM])
+    //     {
+    //        // when a single character check narrows down to one possibility:
+    //        case '1': if (key == "caseM1") { goto labelM1; } else { goto labelDefault; }
+    //
+    //        // when a single character check leaves a few possibilities remaining (worst case scenario):
+    //        case '2':
+    //          switch (key)
+    //          {
+    //            case "caseM1_A": goto labelM1_A;
+    //            case "caseM1_B": goto labelM1_B;
+    //            ...
+    //            default: goto labelDefault;
+    //          }
+    //        ...
+    //        default: goto labelDefault;
+    //      }
+    //    ...
+    //    default: goto labelDefault;
+    // }
+
+    internal class LengthBasedStringSwitchData
+    {
+        internal readonly LengthJumpTable _lengthJumpTable;
+        internal readonly ImmutableArray<CharJumpTable> _charJumpTables;
+        internal readonly ImmutableArray<StringJumpTable> _stringJumpTables;
+
+        internal LengthBasedStringSwitchData(LengthJumpTable lengthJumpTable,
+            ImmutableArray<CharJumpTable> charJumpTables, ImmutableArray<StringJumpTable> stringJumpTables)
+        {
+            _lengthJumpTable = lengthJumpTable;
+            _charJumpTables = charJumpTables;
+            _stringJumpTables = stringJumpTables;
+        }
+
+        internal struct LengthJumpTable
+        {
+            public readonly LabelSymbol? nullCaseLabel;
+            public ImmutableArray<(ConstantValue value, LabelSymbol label)> lengthCaseLabels;
+
+            public LengthJumpTable(LabelSymbol? nullCaseLabel, ImmutableArray<(ConstantValue value, LabelSymbol label)> lengthCaseLabels)
+            {
+                Debug.Assert(lengthCaseLabels.All(c => c.value.IsIntegral) && lengthCaseLabels.Length > 0);
```

Comment:
>IsIntegral [](http://example.com/codeflow?start=63&length=10)

Could we strengthen this to `Discriminator == ConstantValueDiscriminator.Int32`?